data/bigquery/v0/main.go

```diff
@@ -145,3 +149,137 @@ func (c *component) Test(sysVars map[string]any, setup *structpb.Struct) error {
 	}
 	return errors.New("project ID does not match")
 }
+
+type TableColumns struct {
+	TableName string
+	Columns   []Column
+}
+
+type Column struct {
+	Name string
+	Type string
+}
+
+func (c *component) GetDefinition(sysVars map[string]any, compConfig *base.ComponentConfig) (*pb.ComponentDefinition, error) {
+
+	ctx := context.Background()
+	oriDef, err := c.Component.GetDefinition(nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if sysVars == nil && compConfig == nil {
+		return oriDef, nil
+	}
+
+	def := proto.Clone(oriDef).(*pb.ComponentDefinition)
+	client, err := NewClient(compConfig.Setup["json_key"].(string), compConfig.Setup["project_id"].(string))
+	if err != nil || client == nil {
+		return nil, fmt.Errorf("error creating BigQuery client: %v", err)
+	}
+	defer client.Close()
+
+	myDataset := client.Dataset(compConfig.Setup["dataset_id"].(string))
+	tables, err := constructTableColumns(myDataset, ctx, compConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	tableProperties, err := constructTableProperties(tables)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: chuang8511, remove table from definition.json and make it dynamic.
+	tableProperty := tableProperties[0]
+	for _, sch := range def.Spec.ComponentSpecification.Fields["oneOf"].GetListValue().Values {
+		data := sch.GetStructValue().Fields["properties"].GetStructValue().Fields["input"].GetStructValue().Fields["properties"].GetStructValue().Fields["data"].GetStructValue()
+		if data != nil {
+			data.Fields["properties"] = structpb.NewStructValue(tableProperty)
+		}
+	}
+
+	for _, dataSpec := range def.Spec.DataSpecifications {
+		dataInput := dataSpec.Input.Fields["properties"].GetStructValue().Fields["data"].GetStructValue()
+		if dataInput != nil {
+			dataInput.Fields["properties"] = structpb.NewStructValue(tableProperty)
+		}
+		dataOutput := dataSpec.Output.Fields["properties"].GetStructValue().Fields["data"].GetStructValue()
+
+		if dataOutput != nil {
+			aPieceData := dataOutput.Fields["items"].GetStructValue()
+			if aPieceData != nil {
+				aPieceData.Fields["properties"] = structpb.NewStructValue(tableProperty)
+			}
+
+		}
+	}
+
+	return def, nil
+}
+
+func constructTableColumns(myDataset *bigquery.Dataset, ctx context.Context, compConfig *base.ComponentConfig) ([]TableColumns, error) {
+	tableIT := myDataset.Tables(ctx)
+	tables := []TableColumns{}
```

Comment:
How about we only fetch the table we want to use? 

If we still want to fetch multiple tables, we can use a map with table_name as the key.

Reply:
@donch1989 
I was thinking to take the table name from setting parts to input.
Then, the users can choose the table / tables they want to extract with smart hint or drag-and-drop menu.
And, we also can adjust the columns without saving pipeline again if users want to change table.

What do you think about that?

Reply:
Yeah, sounds good

Reply:
I have created another ticket and leave TODO memo here.
I will deal with it in the near future.